### PR TITLE
Fix: A queuing guest can ignore the next guest in the queue in certain circumstances (RCT2 bug)

### LINKED ISSUE #17814: Guests entering an over-capacity queue and "lining up twice"
### Operating System

Windows 10

### OpenRCT2 build

OpenRCT2, v0.4.1-185-g53f7a20 (53f7a20 on develop) provided by GitHub

### Describe the issue

Somehow, this queue line goes over-capacity.  See screenshot, based on this queue length, I would expect ~60 guests, but you can see the queue has 159 people per the Cape Canaveral ride window, and visually, there are a lot of overlapping guests stacked on top of each other in the overworld.  See how at the front of the line, there's no spacing at all between guests because they are overlapping on top of each other, whereas in the pack the spacing is more of what is expected.

I have noticed that even though the line is full, guests are still entering the line, and that they are going all the way to the front, as if ignoring the existing queued guests entirely.

The impact of this is that the queue will exponentially grow, leading to very long queue times longer than the ride capacity, leading to unhappy guests that are waiting in line too long since the queue is over designed capacity.

image

### Area(s) with issue?

_No response_

### Steps to reproduce

1. Load the attached Ivory Towers save file
2. Go to Cape Canaveral, and you will see that guests enter the queue even though the queue is full, and they walk directly to the front of the queue

### Attachments

[Ivory Towers.zip](https://github.com/OpenRCT2/OpenRCT2/files/9332435/Ivory.Towers.zip)

### PR DESCRIPTION
**Pull request by Wirlie**
After some research I think that I have found the issue that I have seen in the original game (and also present in OpenRCT2) while testing my previous PR, this is related to the logic to update the queue position of a guest. 

----
**Explanation:**
The logic to update the queue position of a guest uses the direction of the next guest in the queue to determine if there is space available to walk and also to check if the current guest is still before the next guest in the queue, and according to what I understand, this logic uses the `sprite_direction` to determine if both guests have the same direction:
https://github.com/OpenRCT2/OpenRCT2/blob/8495dbf6a5d7441911f392010aebe3d308b2a8fc/src/openrct2/entity/Guest.cpp#L7328-L7331

If both guests have the same direction, then the next part of the code is used to ensure that the current guest is always before the next guest in the queue:
https://github.com/OpenRCT2/OpenRCT2/blob/8495dbf6a5d7441911f392010aebe3d308b2a8fc/src/openrct2/entity/Guest.cpp#L7332-L7350

The issue here is, when the next guest is looking at another direction different from the current guest, and the next guest is not too nearby to the current guest, the condition `sprite_direction != guestNext->sprite_direction` will be true and method will return with false, allowing the current guest to walk ignoring the next guest in the queue because the logic to ensure that a guest is always before to the next guest in the queue will not be reached.

<details>
  <summary>Circumstances needed  to reproduce this</summary>
<p>

image

Gif:
openrct2_tFCKei0M5P
</p>
</details>

----

**How I fixed it:**
I have used `PeepDirection` to compare directions between both guests instead of using `sprite_direction`

I do not know much about the game logic and I may be wrong, but after digging some methods used by the AI, I have found the method `CalculateNextDestination` in `GuestPathfinding.cpp` and as I understood, this calculates the next queue tile that guest should move in, and after that `peep_move_one_tile` is called with the direction that guest should move to the next tile.

The important thing here is inside the `peep_move_one_tile` method, this set the `PeepDirection` with the direction that guest is currently following.
https://github.com/OpenRCT2/OpenRCT2/blob/8495dbf6a5d7441911f392010aebe3d308b2a8fc/src/openrct2/peep/GuestPathfinding.cpp#L137-L138

~~I think that this variable should always have the real direction that guest is following, and it can be used to check the direction of a guest.~~
**Edit:** After more research, I am now  sure that `PeepDirection` always have the real direction that guest is currently following.

----

I have made some tests and confirmed that using `PeepDirection` fixes the issue.

openrct2_K6VnJnQICn

You can use this save game to reproduce this:
[bug-guest-ignore-next-guest.zip](https://github.com/OpenRCT2/OpenRCT2/files/10363720/bug-guest-ignore-next-guest.zip)

----

This partially fixes #17814 , but there is some remaining things that should be fixed and I will make another PR later to try to address remaining things.

If something is needed to be changed please let me know, thank you!

### CODE REVIEW

File: src/openrct2/entity/Peep.cpp [hunk 1]
```diff
@@ -2236,6 +2236,10 @@ static void peep_interact_with_path(Peep* peep, const CoordsXYE& coords)
                         }
                     }
 
+                    auto queueTileCentre = CoordsXY{ CoordsXY{ guest->NextLoc } + CoordsDirectionDelta[guest->PeepDirection] }
+                                               .ToTileCentre();
+                    guest->SetDestination(queueTileCentre);
+
```
Comment (Wirlie) [inline src/openrct2/entity/Peep.cpp:2242]: I have compared RCT2 and OpenRCT2 and I have see that in RCT2 guests are aligned to the center of the tile after joining to the queue, but in OpenRCT2 guests are aligned to the center of the tile at the second tile and not starting from the first tile.

Also without this, guests can still ignore the next guest in the queue if you configure a queue in this way (tested with maximum game speed):
image

------

After:
image

Before:
image

Comment (Gymnasiast) [reply src/openrct2/entity/Peep.cpp:2242]: This _does_ direct the guest the middle of the tile, in both x and y direction. But if the first queue tile has peeps standing on it, that location will be occupied by other. I don’t know if this is a problem?

Comment (Wirlie) [reply src/openrct2/entity/Peep.cpp:2242]: I do not think that it should be a problem, the logic to update the queue position of a guest takes care about the distance of the next guest in the queue: guest will wait if the distance to the next guest in both axis (X,Y) is less than 7 units no matters if both guests does not have the same direction, I think that this should not interfere with the guest AI (please see my explanation below).

I have done more research and I think that now I have the answer of why guests should be aligned to the middle of the first queue tile.

Sorry for my long text, but I prefer to explain all of my research in order to get more feedback if this issue can be solved in another way or if I have forgot something.

----
When a guest is walking, every time that guest reaches his destination, pathfinding is used to calculate the next tile that guest should walk (it is done by the `CalculateNextDestination()` method at `GuestPathfinding.cpp`) and the destination and direction of the guest is changed to the next tile.

`CalculateNextDestination()` always ends calling `peep_move_one_tile()` and `peep_move_one_tile()` will set the destination of the guest to the middle of the next tile, also `peep_move_one_tile()` will add an offset to the destination to make a spread effect along walking guests that are not queuing.

When a guest joins a queue, his destination will remain having the spread offset because at this point the guest has not yet reached his destination (remember that `CalculateNextDestination()` is only called when a guest reaches his destination), when guest finally reaches his destination pathfinding will change the destination of the guest to the middle of the second queue tile (because guest have the queuing state). [also this is the reason of why guests are aligned to the middle of the second queue tile instead of doing it at the first queue tile]

image

The spread offset added by `peep_move_one_tile()` is done within a square from the middle of the tile
image

----
#### Reason of why guests should always be aligned to the middle of the first queue tile:
If you configures a queue in this way:

image

There is a low chance that a guest can ignore the next guest in queue if certain conditions are meet:
image

Demonstration in-game, the next guest of brown guest is the purple guest:
image

These conditions cannot occurrs if guests are aligned to the middle of the first queue tile because both guests will have the same destination, and due to that, guest will not be able to reach the destination before the next guest in the queue because as I mentioned before, guest will wait if is too nearby to the next guest in the queue.

Thank you for reading this! I hope that this explanation can provide more context of the issue.

**Comment by Wirlie** [comment]
Rebased and tested with latest changes from develop, @Gymnasiast this PR is ready for review,  I do not know what members of the dev team are more familiar with the AI logic but I tagged you, can you please help me reviewing this PR and also requesting a review from another member of the dev team if necessary? Thank you!

Also this PR needs a replay update but I do not know how I can achieve this, I will appreciate if someone can help me with this point.

File: src/openrct2/entity/Guest.cpp [hunk 2]
```diff
@@ -7318,10 +7318,10 @@ bool Guest::UpdateQueuePosition(PeepActionType previous_action)
                 return false;
         }
 
-        if (sprite_direction != guestNext->sprite_direction)
+        if (PeepDirection != guestNext->PeepDirection)
             return false;
 
-        switch (guestNext->sprite_direction / 8)
+        switch (guestNext->PeepDirection)
```
Comment (Gymnasiast) [inline src/openrct2/entity/Guest.cpp:7324]: Could you explain why the `/ 8` was dropped?

Comment (Wirlie) [reply src/openrct2/entity/Guest.cpp:7324]: I'm not too sure about this, but I have see in other parts of the code that `sprite_direction >> 3` is used to set/calculate `PeepDirection`, and `PeepDirection << 3` to set `sprite_direction`. If I am not wrong, `sprite_direction / 8` is equivalent to `sprite_direction >> 3`. I do not know much about the logic of `sprite_direction` but I have see that a walking guest can have a `sprite_direction` of `0`, `8`, `16` and `24` (depending of the direction that guest is looking), and divide it by 8 will result in `0`, `1`, `2` and `3`, comparing this with `PeepDirection` I see that both have the same value:

image

The reason of why I no use `sprite_direction / 8` here is due to the issue that I explained before: a queuing guest can wait in the queue looking at another direction different from the real direction that guest is currently following, and because of that, `sprite_direction` may not always reflect the real direction of a queuing guest. In other hand, `PeepDirection` always have the real direction of a queuing guest.

----
I have made some tests to test the `switch` in all directions.

The purpose of the `switch` is to allow a guest to walk along the queue if the guest is before to the next guest.

To trigger the `switch`, both guests must have a distance greater than 7 units from any axis (X,Y), if distance from both guests is also greater than 13 units then both guest must be on the same tile. Also both guests must have the same direction. 

I have aligned guests to the side of the queue instead of to the middle to accomplish with this condition.

* Test  `#1` using `switch(guestNext->sprite_direction / 8)`:
test_sprite_direction

* Test  `#2` using `switch(guestNext->PeepDirection)` (same result as Test `#1`):
test_peep_direction

    also works for non-straight queue paths:
    test_no_straight_path

* Test  `#3` simulating that the switch is not implemented correctly:
test_switch_fail

I have inverted all cases of the switch to do this test:
```c++
switch(...)
{
   case 3:
      ...
      break;
   case 2:
      ...
      break;
   case 1:
      ...
      break;
   case 0:
      ...
      break;
}
```

* Test  `#4` simulating that the switch is not implemented correctly and none of the switch cases are reached:
test_switch_no_trigger

**Comment by Gymnasiast** [comment]
You said it partially fixes https://github.com/OpenRCT2/OpenRCT2/issues/17814. Which bits does it fix and which bits doesn't it fix?

**Comment by Wirlie** [comment]
The other issue that is not covered by this PR is the fact that guests are able to lining up twice, but I need to do more research to understand why guests are able to lining up twice and to find a way to properly solve this.

This can easily be reproduced if you edit a queue in this way:
demo_line_up_twice

This PR fixes the issue where guests are able to ignore the next guest in the queue if certaim conditions are met without player intervention, allowing them to line up twice, and by fixing this issue, the only way that guests will be able to do this is by the player intervention, when a player edits an existing queue line with guests inside.

In short, the changes made in this PR will ensure that the AI logic related to queuing guests will never fail in normal conditions (without player intervention), making it consistent with the game.

File: src/openrct2/entity/Peep.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/entity/Peep.cpp:2243]: This is some excellent work. Could you provide a super brief summary in a comment in the code at this point. Something like:

```
// Force set centre of tile to prevent issues with guests accidentally skipping the queue
```

File: src/openrct2/entity/Guest.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/openrct2/entity/Guest.cpp:7324]: PeepDirection is the direction that the peep is travelling on the tile. sprite_direction is used for the actual sprite that is drawn. sprite_direction has much higher granularity as it is also used for vehicles but for peeps it is always just * 8, <<3 of the cardinal directions. So this change does make sense I think.

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
Could you add the comment mentioned and bump the network version in NetworkBase.cpp. I'll have to rerecord replays to get the tests to pass.

**Comment by Wirlie** [comment]
Sure, thank you @duncanspumpkin 🙂 . I have increased the network version and added the comment.

**Review by duncanspumpkin — APPROVED** [review]

**Comment by tupaschoal** [comment]
I just got round to reading the comments and investigation here today, but it is great work!
